src/main/java/com/es/phoneshop/service/impl/CartServiceImplement.java

```diff
@@ -78,15 +67,78 @@ public void add(HttpSession session, Long productId, int quantity) {
                 cart.getItems().add(new CartItem(product, quantity));
             }
 
+            recalculateCart(cart);
+
         } finally {
             sessionLock.unlock();
         }
     }
 
+    @Override
+    public void update(HttpSession session, Long productId, int quantity) {
+        Lock sessionLock = SessionLockManager.getSessionLock(session);
+        sessionLock.lock();
+
+        try {
+            Cart cart = getCart(session);
+
+            Product product = productService.getProduct(productId);
+
+            CartItem cartItem = findCartItemByProduct(cart, product);
+
+            validateStock(product, quantity);
+
+            cartItem.setQuantity(quantity);
+
+            recalculateCart(cart);
+
+            session.setAttribute(SESSION_ATTRIBUTE, cart);
```

Comment:
Is there any reason to set cart again to session as it's already there? + same in delete method